### PR TITLE
[jit] user defined types

### PR DESCRIPTION
**Pull request by suo**
First pass at user defined types. The following is contained in this PR:
- `UserType` type, which contains a reference to a module with all methods for the type, and a separate namespace for data attributes (map of name -> TypePtr). 
- `UserTypeRegistry`, similar to the operator registry
- `UserObject` which is the runtime representation of the user type (just a map of names -> IValues)
- `UserTypeValue` SugaredValue, to manage getattr and setattr while generating IR, plus compiler.cpp changes to make that work.
- Frontend changes to get `@torch.jit.script` to work as a class decorator
- `ClassDef` node in our AST.
- primitive ops for object creation, setattr, and getattr, plus alias analysis changes to make mutation safe.

Things that definitely need to get done:
- Import/export, python_print support
- String frontend doesn't understand class definitions yet
- Python interop (using a user-defined type outside TorchScript) is completely broken
- Static methods (without `self`) don't work

Things that are nice but not essential:
- Method definition shouldn't matter (right now you can only reference a method that's already been defined)
- Class definitions can only contain defs, no other expressions are supported.

Things I definitely won't do initially:
- Polymorphism/inheritance

### CODE REVIEW

File: torch/csrc/jit/script/sugared_value.h [hunk 1]
```diff
@@ -157,12 +167,48 @@ struct TORCH_API BuiltinModule : public SugaredValue {
   c10::optional<int64_t> version;
 };
 
+struct TORCH_API UserTypeValue : public SugaredValue {
```
Comment (suo) [inline torch/csrc/jit/script/sugared_value.h]: Right now this is doing double duty as a type and in instance of that type. I plan to separate those concepts

File: torch/csrc/jit/script/sugared_value.cpp [hunk 2]
```diff
@@ -130,6 +130,80 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
...
+    const SourceRange& loc,
+    Method& m,
+    const std::string& field,
+    Value* newValue) {
+  auto expectedType = type_->getAttribute(field);
+  if (!expectedType) {
+    // We don't have an attribute with this name, either add it to the type
+    // definition or throw an error
+    if (m.name() == "__init__") {
+      type_->addAttribute(field, newValue->type());
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: probably also want to check that the current insertion point is not in a sub-block, indicating that this would be a conditional assignment.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 3]
```diff
@@ -130,6 +130,80 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
...
+    throw ErrorReport(loc)
+        << "Tried to access to nonexistent attribute " << field
+        << ". Did you forget to initialize it in __init__()?";
+  }
+  auto& g = *m.graph();
+  auto n = g.insertNode(g.createGetAttr(value_, field));
+  return std::make_shared<SimpleValue>(n->output());
+}
+
+std::shared_ptr<SugaredValue> UserTypeValue::call(
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: This doesn't seem right, this would do:

```
a = <some user type>
b = a() # constructs a new instance of the type of a?
```

Instead we probably want to make a different SugaredValue that represents the class itself which has this logic in it.

File: torch/csrc/jit/script/sugared_value.h [hunk 4]
```diff
@@ -157,12 +167,50 @@ struct TORCH_API BuiltinModule : public SugaredValue {
   c10::optional<int64_t> version;
 };
 
+// TODO right now this is doing double duty as a type and an instance of that
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: oh, yes. I read that out of order.

File: torch/csrc/jit/script/sugared_value.h [hunk 5]
```diff
@@ -172,12 +220,19 @@ struct MethodValue : public SugaredValue {
       at::ArrayRef<NamedValue> inputs,
       at::ArrayRef<NamedValue> attributes,
       size_t n_binders) override {
+    c10::optional<NamedValue> self;
+    if (auto userType = dynamic_cast<UserTypeValue*>(self_.get())) {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: Isn't it safe to just leave self as the first argument? The emit_call_to self logic is for silly builtin things where the self argument is not first.

File: torch/csrc/jit/script/sugared_value.h [hunk 6]
```diff
@@ -157,12 +167,50 @@ struct TORCH_API BuiltinModule : public SugaredValue {
   c10::optional<int64_t> version;
 };
 
+// TODO right now this is doing double duty as a type and an instance of that
+// type, need to separate
+struct TORCH_API UserTypeValue : public SugaredValue {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: Values are first class but all the attribute logic is being handled in a SugaredValue. This isn't going to work in many cases. For instance, once we conditionally assign a UserType, then it will no longer be a SugaredValue.

It makes sense for UserTypeValue to handle the special logic for `__init__` assignment but otherwise the logic for normal assignment and normal attribute lookup should be defined in SimpleValue. UserTypeValue is probably a subclass of SimpleValue.

File: torch/csrc/jit/script/compiler.cpp [hunk 7]
```diff
@@ -629,8 +631,9 @@ struct to_ir {
       const SugaredValuePtr& self) {
     auto params_begin = decl.params().begin();
     auto params_end = decl.params().end();
-    if (self)
+    if (self && !dynamic_cast<UserTypeValue*>(self.get())) {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: Thought about it more. I don't think there should be a UserDataValue sugared value at all. Assignment and attribute lookup need to work on plain SimpleValues because these objects are first class. The __init__ handling can be special cased into that logic (and it also enables us to do more strict checks, like is it unconditionally assigning or not).

File: aten/src/ATen/core/jit_type.h [hunk 8]
```diff
@@ -1068,4 +1077,79 @@ matchTypeVariables(TypePtr formal, TypePtr actual, TypeEnv& type_env);
...
+      return namespace_.at(name);
+    }
+    return nullptr;
+  }
+
+  std::string name() const {
+    return typename_;
+  }
+
+  std::shared_ptr<Module> module() const {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: More thoughts: 

Should we expand our builtin functions to just hold all methods as well, and use that standard interface to lookup implementations.

Or should we allow _any type_ to have a list of methods/builtin ops, and move all our current builtin ops to these tables.

File: torch/jit/__init__.py [hunk 9]
```diff
@@ -712,18 +713,24 @@ def _try_compile_weak_script(fn):
...
     if not _enabled:
-        return fn
+        return obj
     if _rcb is None:
         _rcb = _jit_internal.createResolutionCallback(_frames_up + 1)
-    ast = get_jit_ast(fn, is_method=False)
-    mod = ScriptModule()
-    _jit_script_compile(mod, ast, _rcb, get_default_args(fn))
+    if inspect.isclass(obj):
+        mod = ScriptClass(obj.__name__)
```
Comment (zdevito) [inline torch/jit/__init__.py]: We should consider namespaces as well. These are going to be globally registered in JIT. We may want to use the fully qualified python module name or allow the user to specify qualified name for a type.

**Review by zdevito — COMMENTED** [review]
This is a great start!  Overall it looks like the right approach and I put a bunch of comments inline with suggestions.

File: torch/csrc/jit/script/compiler.cpp [hunk 10]
```diff
@@ -715,21 +718,37 @@ struct to_ir {
     // inputs
     auto it = def.decl().params().begin();
     auto end = def.decl().params().end();
-    auto expected_annotation_size =
-        self ? def.decl().params().size() - 1 : def.decl().params().size();
+    auto userType = dynamic_cast<UserTypeValue*>(self.get());
+    auto expected_annotation_size = def.decl().params().size();
+    if (self && !userType) {
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/script/compiler.cpp]: `expected_annotation_size` is a misnomer, this is really just checking that the schema arguments match the decl. The JIT schema should have an explicit `self` argument, except for when self is a module/not first class.

File: torch/jit/__init__.py (hunk 9 above)
Comment (suo) [reply torch/jit/__init__.py]: Let's leave this for a follow-up PR. It's a little tricky, since right now a user may want to annotate a type `Foo` but we have no way of going from that name to a fully-qualified type without pulling a python dependency into the type parser.

File: aten/src/ATen/core/ivalue.h [hunk 11]
```diff
@@ -91,7 +91,39 @@ using DoubleList = List<double>;
 using BoolList = List<bool>;
 using GenericList = List<IValue>;
 
+// User-defined object.
+template <typename Elem>
+struct CAFFE2_API UserObjectGeneric final : c10::intrusive_ptr_target {
+ public:
+  UserObjectGeneric(std::string name, size_t numSlots)
```
Comment (apaszke) [inline aten/src/ATen/core/ivalue.h]: Can't we use interned strings for the name? Right now every instance of an object will keep a separately allocated `std::string`. Seems really wasteful

File: aten/src/ATen/core/jit_type.h [hunk 12]
```diff
@@ -1068,4 +1079,111 @@ matchTypeVariables(TypePtr formal, TypePtr actual, TypeEnv& type_env);
...
+  }
+
+  size_t numAttributes() const {
+    return attributes_.size();
+  }
+
+  // Attributes are stored in a specific slot at runtime for effiency.
+  // When emitting instructions we specify the slot so that attribute access is
+  // a constant lookup
+  size_t getAttributeSlot(const std::string& name) const {
```
Comment (apaszke) [inline aten/src/ATen/core/jit_type.h]: I feel like both the type names and attribute names could get interned safely. Do we really feel like people will be creating lots of those dynamically?

File: test/test_jit.py [hunk 13]
```diff
@@ -13397,6 +13397,114 @@ def test_tensor_sharing(self):
...
+
+    def test_type_annotations(self):
+        # Remove this when import/export is implemented for classes
+        with self.disableModuleHook():
+            with self.assertRaisesRegex(RuntimeError, "expected a value of type bool"):
+                @torch.jit.script
+                class Foo:
+                    def __init__(self, x):
+                        # type: (bool)
+                        self.foo = x
```
Comment (apaszke) [inline test/test_jit.py]: Can we get some tests where a `script`-decorated class is used directly from Python?

File: tools/clang_format.py [hunk 14]
```diff
@@ -23,6 +21,9 @@
 CLANG_FORMAT_WHITELIST = ["torch/csrc/jit/", "test/cpp/jit/"]
 
 CPP_FILE_REGEX = re.compile("^.*\\.(h|cpp|cc|c|hpp)$")
+CPP_FILE_REGEX = re.compile(".*\\.(h|cpp|cc|c|hpp)$")
```
Comment (apaszke) [inline tools/clang_format.py]: Can we put this in a separate patch please?

File: aten/src/ATen/core/ivalue.h (hunk 11 above)
Comment (apaszke) [reply aten/src/ATen/core/ivalue.h]: Or at least make this a `shared_ptr`

File: tools/clang_format.py (hunk 14 above)
Comment (suo) [reply tools/clang_format.py]: oops this got included in somehow, will remove

File: test/test_jit.py (hunk 13 above)
Comment (suo) [reply test/test_jit.py]: Right now that doesn't work :(. But forthcoming!

File: aten/src/ATen/core/ivalue.h (hunk 11 above)
Comment (suo) [reply aten/src/ATen/core/ivalue.h]: good catch, will change

File: aten/src/ATen/core/jit_type.h (hunk 12 above)
Comment (suo) [reply aten/src/ATen/core/jit_type.h]: When we are emitting getattr and setattr instructions we're getting the names in string form from the AST anyway, so the interface needs to accept strings at least. I think it's okay not to intern since it's only one string per attribute per type.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 15]
```diff
@@ -816,7 +816,39 @@ RegisterOperators reg({
...
+             auto value = userObj->getAttr(slot);
+             push(stack, std::move(value));
+             return 0;
+           };
+         }),
+     Operator(prim::SetAttr, [](const Node* node) {
+       return [node](Stack& stack) {
+         auto v = pop(stack);
+         auto userObj = pop(stack).toUserObject();
+         const auto slot = node->i(attr::slot);
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: similar here

File: torch/csrc/jit/script/sugared_value.cpp [hunk 16]
```diff
@@ -125,22 +141,91 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
...
+  throw ErrorReport(loc) << value_->type()->str()
                          << " cannot be used as a tuple";
 }
+
+void SimpleValue::setAttr(
+    const SourceRange& loc,
+    Method& m,
+    const std::string& field,
+    Value* newValue,
+    bool shouldDefine) {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: `shouldDefine` is _very_ specific to Classes and not something I would expect to appear in the `setAttr` signature. Another way to organize this would be:

* `setAttr/getAttr` are defined on SimpleValue but never change the members of class.
* A sugared value, `InitializingClassValue` is used for `self` _only for the the init method_, and contains extra logic on setAttr to do the creation.

**Review by zdevito — APPROVED** [review]
Looks good for first patch, assuming renaming UserType -> Class.  

Incomplete list of follow up fixes for further patches:

* Have a different SugaredValue for representing the class object and the instance object
* Namespaces

File: torch/csrc/jit/script/sugared_value.cpp (hunk 16 above)
Comment (suo) [reply torch/csrc/jit/script/sugared_value.cpp]: agree. Will follow up in the next PR with this.
